Fix memory leak and error handling of workfiles.

hlinnaka opened the pull request:
There's no need to allocate ExecWorkFile struct in TopMemoryContext. We
don't rely on a callback to clean them up anymore, they will be closed
by the normal ResourceOwner and MemoryContext mechanisms. Otherwise,
the memory allocated for the struct is leaked on abort.

In the passing, make the error handling more robust.

Per report in private from Karthikeyan Rajaraman.

hlinnaka commented:
This fixes the memory leak reported in private Pivotal tracker story https://www.pivotaltracker.com/story/show/136267837

karthijrk commented on src/backend/executor/execWorkfile.c:
```diff
@@ -45,8 +45,8 @@ ExecWorkFile_Create(const char *fileName,
 {
-	ExecWorkFile *workfile = NULL;
-	void *file = NULL;
+	ExecWorkFile *workfile;
```
Why do we leave the variable uninitialized while declaring it ?

kuien replied:
because, by this, compiler may detect whether the variable is used later and throw warning if necessary.

danielgustafsson replied:
This is to catch a completely different class of errors, the exact opposite actually.

`-Wunused-variable` warns if a variable if unused throughout it's scope, we have this turned on via `-Wall`. This a helpful warning for a trivial, and generally harmless, problem.

By not initalizing the variable at declaration we however get warnings when the variable is indeed used but is read before written to. This is clearly unintentional and a bug in the codepath, but if the variable is initialized at declaration it will go undetected which can lead to subtle and hard to find bugs. Coverity will consider potential reads before writes a defect which has helped us find bugs in the past.

karthijrk commented on src/backend/executor/execWorkfile.c:
```diff
@@ -55,28 +55,25 @@ ExecWorkFile_Create(const char *fileName,
-	 * is still available when calling the transaction callback at the
-	 * time when the transaction aborts.
-	 */
-	 MemoryContext oldContext = MemoryContextSwitchTo(TopMemoryContext);
```
I can't think of a reason why we would need to use `TopMemoryContext`. But in order to do these changes, I think we would need the following.

1.  ICG doesn't have enough test to test all the caller of `ExecWorkfile`. Before we do any more refactoring, adding a test to exercise that code is a must.
2. I think we still use XCallBackOnce for `ntuplestore_cleanup`. So I wonder if this `ExecWorkfile` exist when this callback gets called.
3. For some reason, even the called for `ExecWorkfile_Create` is changing the context to `TopMemoryContext`. It would be great if we can get rid of this.


@armenatzoglou @foyzur @hardikar Please add your comments if you can think of some issue with this.

karthijrk commented on src/backend/executor/execWorkfile.c:
```diff
@@ -55,28 +55,25 @@ ExecWorkFile_Create(const char *fileName,
+			if (!file)
+				ereport(ERROR,
+						(errcode_for_file_access(),
+						 errmsg("could not create temporary file \"%s\": %m",
```
What is this `%m` for ?

karthijrk replied:
I see there is no similar check for `bfz`. How about having this check after the `switch` case so we cover for `bfz` as well.

karthijrk commented on src/backend/executor/test/execWorkfile_test.c:
```diff
@@ -1,149 +0,0 @@
-#include <stdarg.h>
```
There is a separate test to just verify that `ExecWorkfile` is created in `TopMemoryContext`. Do you know why there is such a test?

hlinnaka commented:
On 12/20/2016 02:48 PM, Karthikeyan Jambu Rajaraman wrote:
> karthijrk requested changes on this pull request.
>
> In general, thank a lot Heikki for working on this leak issue. Appreciate your effort :)
>
>
>
>> @@ -45,8 +45,8 @@ ExecWorkFile_Create(const char *fileName,
>  					bool delOnClose,
>  					int compressType)
>  {
> -	ExecWorkFile *workfile = NULL;
> -	void *file = NULL;
> +	ExecWorkFile *workfile;
>
> Why do we leave the variable uninitialized while declaring it ?

It's better to not initialize a variable, unless you actually need it to 
be initialized. That allows the compiler to give a warning, if you 
accidentally use the variable before setting it to a proper value. It's 
also more readable to humans, as you don't need to wonder what the NULL 
value might be used for.

It compiles to the same result in this case, but it's just a good habit 
in general.

>> @@ -55,28 +55,25 @@ ExecWorkFile_Create(const char *fileName,
>  		workfile_mgr_report_error();
>  	}
>
> -	/*
> -	 * Create ExecWorkFile in the TopMemoryContext since this memory context
> -	 * is still available when calling the transaction callback at the
> -	 * time when the transaction aborts.
> -	 */
> -	 MemoryContext oldContext = MemoryContextSwitchTo(TopMemoryContext);
>
> I can't think of a reason why we would need to use `TopMemoryContext`. But in order to do these changes, I think we would need the following.
>
> 1.  ICG doesn't have enough test to test all the caller of `ExecWorkfile`. Before we do any more refactoring, adding a test to exercise that code is a must.

Ok. Do we have tests in the TINC suite that you're going to port, or do 
we need to write whole new tests? What codepaths do you think needs more 
coverage?

> 2. I think we still use XCallBackOnce for `ntuplestore_cleanup`. So I wonder if this `ExecWorkfile` exist when this callback gets called.

Oh, ouch, ntuplestorenew.c looks scary :-(. The memory allocation is 
weird, it uses palloc() for some things, but malloc() for others. Sooner 
or later, we need to switch to using the regular tuplestore, and throw 
away ntuplestorenew. Ntuplestore supportes multiple readers, but 
PostgreSQL's tuplestore got support for that in 8.4 as well.

But answer your question, the end-of-xact callback that ntuplestore 
uses, is called before the ResourceOwnerRelease, in AbortTransaction. So 
I think we're good in that regard. I don't have much confidence in the 
code in general, but I guess it works because we're not hearing 
complaints from users.

> 3. For some reason, even the called for `ExecWorkfile_Create` is changing the context to `TopMemoryContext`. It would be great if we can get rid of this.

Yeah.

> @armenatzoglou @foyzur @hardikar Please add your comments if you can think of some issue with this.
>
>
>
>>  	switch(fileType)
>  	{
>  		case BUFFILE:
>  			file = (void *) BufFileCreateFile(fileName, delOnClose, false /* interXact */ );
> +			if (!file)
> +				ereport(ERROR,
> +						(errcode_for_file_access(),
> +						 errmsg("could not create temporary file \"%s\": %m",
>
> What is this `%m` for ?

It prints an error message based on 'errno'. It's used in pretty much 
all error messages after standard library calls, that set errno.

Hmm, looking at BufFileCreateFile(), I don't think it can actually 
return NULL. So this whole check is unnecessary, after all.

>>  	switch(fileType)
>  	{
>  		case BUFFILE:
>  			file = (void *) BufFileCreateFile(fileName, delOnClose, false /* interXact */ );
> +			if (!file)
> +				ereport(ERROR,
> +						(errcode_for_file_access(),
> +						 errmsg("could not create temporary file \"%s\": %m",
>
> I see there is no similar check for `bfz`. How about having this check after the `switch` case so we cover for `bfz` as well.

Same here, BufFileCreateFile() doesn't actually ever return NULL. I'll 
go and remove the checks, and add a comment to BufFileCreateFile() to 
state that explicitly.

>> @@ -1,149 +0,0 @@
> -#include <stdarg.h>
>
> There is a separate test to just verify that `ExecWorkfile` is created in `TopMemoryContext`. Do you know why there is such a test?

Not sure, it seems excessive to me. It was added back in 2014, by commit 
01bdcb02cbba16c0ec21fbf8367cfa246074249e. The commit message said:

>      MPP-21425
>      Allocating ExecWorkFile in TopMemoryContext so that it can be cleaned up during AbortXact
>      CR-9040
>      CL (fix): 182551
>      CL (test): 182552

So it was added as part of attempting to handle abort correctly. But we 
have a better mechanism for that now.

- Heikki

karthijrk commented:
On Tue, Dec 20, 2016 at 5:36 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> On 12/20/2016 02:48 PM, Karthikeyan Jambu Rajaraman wrote:
> > karthijrk requested changes on this pull request.
> >
> > In general, thank a lot Heikki for working on this leak issue.
> Appreciate your effort :)
> >
> >
> >
> >> @@ -45,8 +45,8 @@ ExecWorkFile_Create(const char *fileName,
> > bool delOnClose,
> > int compressType)
> > {
> > - ExecWorkFile *workfile = NULL;
> > - void *file = NULL;
> > + ExecWorkFile *workfile;
> >
> > Why do we leave the variable uninitialized while declaring it ?
>
> It's better to not initialize a variable, unless you actually need it to
> be initialized. That allows the compiler to give a warning, if you
> accidentally use the variable before setting it to a proper value. It's
> also more readable to humans, as you don't need to wonder what the NULL
> value might be used for.
>
> It compiles to the same result in this case, but it's just a good habit
> in general.
>

Cool. Sounds good.

>
> >> @@ -55,28 +55,25 @@ ExecWorkFile_Create(const char *fileName,
> > workfile_mgr_report_error();
> > }
> >
> > - /*
> > - * Create ExecWorkFile in the TopMemoryContext since this memory context
> > - * is still available when calling the transaction callback at the
> > - * time when the transaction aborts.
> > - */
> > - MemoryContext oldContext = MemoryContextSwitchTo(TopMemoryContext);
> >
> > I can't think of a reason why we would need to use `TopMemoryContext`.
> But in order to do these changes, I think we would need the following.
> >
> > 1. ICG doesn't have enough test to test all the caller of
> `ExecWorkfile`. Before we do any more refactoring, adding a test to
> exercise that code is a must.
>
> Ok. Do we have tests in the TINC suite that you're going to port, or do
> we need to write whole new tests? What codepaths do you think needs more
> coverage?
>

I don't think we have any test from tinc. I may be wrong. But we have some
sql files that pretty much excercise different workfile caller.  I can
forward you the email.

>
> > 2. I think we still use XCallBackOnce for `ntuplestore_cleanup`. So I
> wonder if this `ExecWorkfile` exist when this callback gets called.
>
> Oh, ouch, ntuplestorenew.c looks scary :-(. The memory allocation is
> weird, it uses palloc() for some things, but malloc() for others. Sooner
> or later, we need to switch to using the regular tuplestore, and throw
> away ntuplestorenew. Ntuplestore supportes multiple readers, but
> PostgreSQL's tuplestore got support for that in 8.4 as well.
>
> But answer your question, the end-of-xact callback that ntuplestore
> uses, is called before the ResourceOwnerRelease, in AbortTransaction. So
> I think we're good in that regard. I don't have much confidence in the
> code in general, but I guess it works because we're not hearing
> complaints from users.
>
> I assume all the MemoryContext and resource goes aways in
`PrepareTransaction`. But XcallbackOnce for some reason doesn't call
callbacks (`ntuplestore_cleanup`) during `PrepareTransaction`. This
callback will be called in `CommitTransaction` but by then all context
would have gone. In that case, that might be an issue.

I don't have much background on this code as well. I wonder if we need this
callback in first place if all the memory are allocated in appropriate
memory context.


> > 3. For some reason, even the called for `ExecWorkfile_Create` is
> changing the context to `TopMemoryContext`. It would be great if we can get
> rid of this.
>
> Yeah.
>
> > @armenatzoglou @foyzur @hardikar Please add your comments if you can
> think of some issue with this.
> >
> >
> >
> >> switch(fileType)
> > {
> > case BUFFILE:
> > file = (void *) BufFileCreateFile(fileName, delOnClose, false /*
> interXact */ );
> > + if (!file)
> > + ereport(ERROR,
> > + (errcode_for_file_access(),
> > + errmsg("could not create temporary file \"%s\": %m",
> >
> > What is this `%m` for ?
>
> It prints an error message based on 'errno'. It's used in pretty much
> all error messages after standard library calls, that set errno.
>
> Hmm, looking at BufFileCreateFile(), I don't think it can actually
> return NULL. So this whole check is unnecessary, after all.
>

Sounds good.

>
> >> switch(fileType)
> > {
> > case BUFFILE:
> > file = (void *) BufFileCreateFile(fileName, delOnClose, false /*
> interXact */ );
> > + if (!file)
> > + ereport(ERROR,
> > + (errcode_for_file_access(),
> > + errmsg("could not create temporary file \"%s\": %m",
> >
> > I see there is no similar check for `bfz`. How about having this check
> after the `switch` case so we cover for `bfz` as well.
>
> Same here, BufFileCreateFile() doesn't actually ever return NULL. I'll
> go and remove the checks, and add a comment to BufFileCreateFile() to
> state that explicitly.
>
> sounds good.


> >> @@ -1,149 +0,0 @@
> > -#include <stdarg.h>
> >
> > There is a separate test to just verify that `ExecWorkfile` is created
> in `TopMemoryContext`. Do you know why there is such a test?
>
> Not sure, it seems excessive to me. It was added back in 2014, by commit
> 01bdcb02cbba16c0ec21fbf8367cfa246074249e. The commit message said:
>
> > MPP-21425
> > Allocating ExecWorkFile in TopMemoryContext so that it can be cleaned up
> during AbortXact
> > CR-9040
> > CL (fix): 182551
> > CL (test): 182552
>
> So it was added as part of attempting to handle abort correctly. But we
> have a better mechanism for that now.
>
> Ok.

> - Heikki
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/1464#issuecomment-268244634>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFLlg3Zgx_mGIblzX8htzUZY--ZcQoFXks5rJ9nugaJpZM4LRnRg>
> .
>

hlinnaka commented:
I picked the regression test from PR #1602, thanks @karthijrk! I believe this is ready for commit now.

This doesn't do anything about the TopMemoryContext usage in tuplestorenew.c that @karthijrk pointed out. That should be cleaned up too, but it's a separate and slightly bigger issue, because I think we should modify tuplestorenew.c to get rid of the end-of-xact callback altogether, and have it allocate everything in memory contexts, instead of using gp_malloc().

hlinnaka commented:
On closer look, the new regression tests are a few bricks shy of a load:

* Each test does "drop language if exists plpythonu cascade;". That can drop a function that another one of the concurrent tests just created.
* There's no point in doing "set search_path to materialize_spill; RESET ALL;". The RESET ALL will immediately reset the search_path. Therefore, the test tables were not actually being created in the newly-created schema, as was probably intended.
* No need to reset all the GUC individually at the end of each test. (especially when there's also a RESET ALL at the end, but even without that, each test file runs in a separate session)
* Since the test tables are scaled with the number of segments, the output of the COUNT(*) queries depend on the number of segments. Therefore, they will fail, unless you have exactly three segments in the cluster.
* The materialize_spill test takes almost 2 minutes on my laptop, which is way too long to be included in installcheck-good (unless there's a good reason).
* The materialize_spill test ran almost the same explain analyze query twice, once with a LIMIT and once without. Why? Does the limit somehow affect how workfiles work?

I added a commit to this PR to fix all those issues for the materialize_spill test. But the others need a similar treatment. Please have a look at the changes to materialize_spill, and if those are OK, I'll do the same to the other tests.

karthijrk commented:
Great. Thanks for the review on the test. Changes on the test look good. I thought having limit will set the `QueryFinishPending` but it's not. Otherwise, the query with `limit` has to be removed. 

After carefully going through the code, I see some of the changes in TupleStoreNew is missing. This might cause a crash in some corner cases.

E.g. Assume that query that uses tuplestore's workfile write fails in subtransaction but the top transaction does prepare and commit. When the top transaction goes to `PrepareTransaction`, all the temporary files will be gone. During subsequent `CommitTransaction`, we will call `NTupleStoreCleanup` which in turn call `ExecWorkFile_Close`. But the file is already deleted by `ResourceOwner`. 

It took me a while to find out what is missing when you ported the test. So let's iron out the difference before we port the commits or push it.

hlinnaka commented:
On 01/20/2017 10:24 PM, Karthikeyan Jambu Rajaraman wrote:
> Great. Thanks for the review on the test. Changes on the test look
> good. I thought having limit will set the `QueryFinishPending` but
> it's not. Otherwise, the query with `limit` has to be removed.
>
> After carefully going through the code, I see some of the changes in
> TupleStoreNew is missing. This might cause a crash in some corner
> cases.
>
> E.g. Assume that query that uses tuplestore's workfile write fails in
> subtransaction but the top transaction does prepare and commit. When
> the top transaction goes to `PrepareTransaction`, all the temporary
> files will be gone. During subsequent `CommitTransaction`, we will
> call `NTupleStoreCleanup` which in turn call `ExecWorkFile_Close`.
> But the file is already deleted by `ResourceOwner`.
>
> It took me a while to find out what is missing when you ported the
> test. So let's iron out the difference before we port the commits or
> push it.

Hmm, I see. Did you have a fix for that in PR #1602? I didn't see it.

- Heikki

karthijrk commented:
Yes, the fix is in PR #1602. It should be part of the commit - https://github.com/greenplum-db/gpdb/pull/1602/commits/615e9d4144b021730723a15ece0f38b826d0e35e#diff-b37908413e016b54a45d549cf0539121. 

Also verified that having `Limit` in the query will set the `QueryFinishPending` to true. Hence we though it is good to have those in the test to make sure that query runs fine without crashing for some reason w.r.t workfile. 

Thanks @armenatzoglou for the suggestion.

hlinnaka commented:
On 01/20/2017 11:38 PM, Karthikeyan Jambu Rajaraman wrote:
> Yes, the fix is in PR #1602. It should be part of the commit -
> https://github.com/greenplum-db/gpdb/pull/1602/commits/615e9d4144b021730723a15ece0f38b826d0e35e#diff-b37908413e016b54a45d549cf0539121.

I don't see it. There are two diff hunks to tuplestorenew.c, and neither 
seems to do anything about the issue you mentioned. Can you hit me with 
the cluebat?

> Also verified that having `Limit` in the query will set the
> `QueryFinishPending` to true. Hence we though it is good to have
> those in the test to make sure that query runs fine without crashing
> for some reason w.r.t workfile.

Ok, let's keep it with some comments, then.

- Heikki

hlinnaka commented:
Ok, I went ahead and modified tuplestorenew.c to use MemoryContexts for all the allocations, and got rid of the end-of-xact callback altogether. I modified the remaining test cases the similarly to what I did for materialize_spill earlier. And I put back the LIMIT test into materialize_spill. Rebased over current master, and squashed.

I believe I've addressed all the comments now, but please have one more look.
